### ISSUE TITLE
Race condition bug: Benches are supposed to be serial, but overlap in setup/teardown

### ISSUE THREAD

**Issue by kahoot-karl** [tinylibs/tinybench#179]
# Repro

```js
var { Bench } = require('tinybench');

const bench = new Bench({
  time: 100,
  teardown: async (test) => {
    console.log('starting teardown...', test.name);
    await new Promise((resolve) => setTimeout(resolve, 1000));
    console.log('ending teardown...', test.name);
  },
  setup: async (test) => {
    console.log('starting setup...', test.name);
    await new Promise((resolve) => setTimeout(resolve, 1000));
    console.log('ending setup...', test.name);
  },
});

bench
  .add('switch 1', () => {
// code
  })
  .add('switch 2', () => {
// code
  });

bench.run().then(() =>
  console.table(
    bench.tasks.map(({ name, result }) => ({
      'Task Name': name,
      'Average Time (ps)': result?.mean * 1000,
      'Variance (ps)': result?.variance * 1000,
    }))
  )
);
```
output:
```
starting setup... switch 1
ending setup... switch 1
starting teardown... switch 1
starting setup... switch 2
ending teardown... switch 1
ending setup... switch 2
starting teardown... switch 2
┌─────────┬────────────┬──────────────────────┬─────────────────────────┐
│ (index) │ Task Name  │ Average Time (ps)    │ Variance (ps)           │
├─────────┼────────────┼──────────────────────┼─────────────────────────┤
│ 0       │ 'switch 1' │ 0.030831958177052855 │ 0.000018364921947085314 │
│ 1       │ 'switch 2' │ 0.03035000904434523  │ 0.000013611904045635386 │
└─────────┴────────────┴──────────────────────┴─────────────────────────┘
ending teardown... switch 2
```

# The problem

From output above:

```
starting teardown... switch 1
starting setup... switch 2
ending teardown... switch 1
```

The startup/teardown are interleaving.

---

This might be the root cause of https://github.com/tinylibs/tinybench/issues/71. Also causes issues upstream in vitest

**Comment by kahoot-karl** [comment]
Indeed. Latest version of uses `2.0.3`, so that is probably the issue

**Comment by kahoot-karl** [comment]
Output on `3.0.3`:

```
starting setup... switch 1
ending setup... switch 1
starting teardown... switch 1
ending teardown... switch 1
starting setup... switch 2
ending setup... switch 2
starting teardown... switch 2
ending teardown... switch 2
starting setup... switch 1
ending setup... switch 1
starting teardown... switch 1
ending teardown... switch 1
starting setup... switch 2
ending setup... switch 2
starting teardown... switch 2
ending teardown... switch 2
┌─────────┬────────────┬──────────────────────┬─────────────────────────┐
│ (index) │ Task Name  │ Average Time (ps)    │ Variance (ps)           │
├─────────┼────────────┼──────────────────────┼─────────────────────────┤
│ 0       │ 'switch 1' │ 0.026681479327522562 │ 0.000005211778207989301 │
│ 1       │ 'switch 2' │ 0.026675830684520604 │ 0.00005034150977909062  │
└─────────┴────────────┴──────────────────────┴─────────────────────────┘
```


They are not interleaving, but the cycles are alternating, is this intentional? It could be nice to have some docs on how it is expected to work.

**Comment by jerome-benoit** [comment]
> They are not interleaving, but the cycles are alternating, is this intentional? It could be nice to have some docs on how it is expected to work.

+ warmup run if enabled
+ then real benchmark run

You can choose at which step a block of code should be run at setup and teardown:  https://github.com/tinylibs/tinybench/blob/main/examples/src/simple-gc.ts

**Comment by kahoot-karl** [comment]
Right, the warmup runs all the tests before starting the actual benching, thanks for the explanation!
